iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java

```diff
@@ -1121,39 +1216,57 @@ private void updateDeviceToSourceTransformAndOutputExpressions(
+    }
+
     if (queryStatement.isAggregationQuery()) {
-      for (Expression expression : analysis.getAggregationExpressions()) {
+      for (Expression aggregationExpression : analysis.getAggregationExpressions()) {
+        // do not put "TimestampOperand" into sourceTransformExpression
+        if (aggregationExpression instanceof FunctionExpression
+            && COUNT_TIME.equalsIgnoreCase(
+                ((FunctionExpression) aggregationExpression).getFunctionName())) {
+          continue;
+        }
+
         // for AggregationExpression, only the first Expression of input need to transform
-        sourceTransformExpressions.add(expression.getExpressions().get(0));
+        sourceTransformExpressions.add(aggregationExpression.getExpressions().get(0));
       }
+
       if (queryStatement.hasGroupByExpression()) {
         sourceTransformExpressions.add(analysis.getGroupByExpression());
       }
-    } else {
-      sourceTransformExpressions.addAll(analysis.getSelectExpressions());
-      if (queryStatement.hasOrderByExpression()) {
-        sourceTransformExpressions.addAll(analysis.getOrderByExpressions());
-      }
+
+      return;
+    }
+
+    sourceTransformExpressions.addAll(analysis.getSelectExpressions());
+    if (queryStatement.hasOrderByExpression()) {
+      sourceTransformExpressions.addAll(analysis.getOrderByExpressions());
     }
-    analysis.setSourceTransformExpressions(sourceTransformExpressions);
   }
 
   private void analyzeDeviceToSource(Analysis analysis, QueryStatement queryStatement) {
-    Map<String, Set<Expression>> deviceToSourceExpressions = new HashMap<>();
+    Map<String, Set<Expression>> deviceToSourceExpressions =
+        Optional.ofNullable(analysis.getDeviceToSourceExpressions()).orElse(new HashMap<>());
     Map<String, Set<Expression>> deviceToSourceTransformExpressions =
         analysis.getDeviceToSourceTransformExpressions();
-    for (Map.Entry<String, Set<Expression>> deviceSourceTransformExpressionsEntry :
-        deviceToSourceTransformExpressions.entrySet()) {
-      String deviceName = deviceSourceTransformExpressionsEntry.getKey();
-      Set<Expression> sourceTransformExpressions = deviceSourceTransformExpressionsEntry.getValue();
+
+    for (Map.Entry<String, Set<Expression>> entry : deviceToSourceTransformExpressions.entrySet()) {
+      String deviceName = entry.getKey();
+      Set<Expression> sourceTransformExpressions = entry.getValue();
 
       Set<Expression> sourceExpressions = new LinkedHashSet<>();
       for (Expression expression : sourceTransformExpressions) {
-        sourceExpressions.addAll(ExpressionAnalyzer.searchSourceExpressions(expression));
+        sourceExpressions.addAll(searchSourceExpressions(expression));
       }
-      deviceToSourceExpressions.put(deviceName, sourceExpressions);
+      deviceToSourceExpressions
+          .computeIfAbsent(deviceName, key -> new LinkedHashSet<>())
+          .addAll(sourceExpressions);
```

Comment:
should directly `put` instead of `addAll`

Reply:
fixed